c++/src/kj/async-inl.h

```diff
@@ -1470,4 +1474,413 @@ PromiseForResult<Func, void> Executor::executeAsync(Func&& func) const {
+  KJ_DISALLOW_COPY(Coroutine);
+
+  Promise<T> get_return_object() {
+    // Called after coroutine frame construction and before initial_suspend() to create the
+    // coroutine's return object. `this` itself lives inside the coroutine frame, and we arrange for
+    // the returned Promise<T> to own `this` via a custom Disposer and by always leaving the
+    // coroutine in a suspended state.
+    return PromiseNode::to<Promise<T>>(Own<PromiseNode>(this, *this));
+  }
+
+private:
+  // -------------------------------------------------------
+  // Disposer implementation
+
+  void disposeImpl(void* pointer) const override {
+    KJ_IASSERT(pointer == this);
+
+    // const_cast okay -- every Own<PromiseNode> that we build in get_return_object() uses itself
+    // as the disposer, thus every disposer is unique and there are no thread-safety concerns.
+    const_cast<Coroutine&>(*this).destroy();
+  }
+
+  void destroy() {
+    // Mutable helper function for disposeImpl(). Basically a wrapper around coroutine.destroy()
+    // with some stuff to propagate exceptions appropriately.
+
+    auto coroutine = Handle::from_promise(*this);
+
+    // Objects in the coroutine frame might throw from their destructors, so unhandled_exception()
+    // will need some way to communicate those exceptions back to us. Separately, we also want
+    // confirmation that our own ~Coroutine() destructor ran. To solve this, we put a
+    // DisposalResults object on the stack and set a pointer to it in the Coroutine object. This
+    // indicates to unhandled_exception() and ~Coroutine() where to store the results of the
+    // destruction operation.
+    DisposalResults disposalResults;
+    maybeDisposalResults = &disposalResults;
+
+    // Need to save this while `unwindDetector` is still valid.
+    bool shouldRethrow = !unwindDetector.isUnwinding();
+
+    do {
+      // Clang's implementation of the Coroutines TS does not destroy the Coroutine object or
+      // deallocate the coroutine frame if a destructor of an object on the frame threw an
+      // exception. This is despite the fact that it delivered the exception to _us_ via
+      // unhandled_exception(). Anyway, it appears we can work around this by running
+      // coroutine.destroy() a second time.
+      //
+      // On Clang, `disposalResults.exception != nullptr` implies `!disposalResults.destructorRan`.
+      // We could optimize out the separate `destructorRan` flag if we verify that other compilers
+      // behave the same way.
+      coroutine.destroy();
+    } while (!disposalResults.destructorRan);
+
+    // WARNING: `this` is now a dangling pointer.
+
+    KJ_IF_MAYBE(exception, disposalResults.exception) {
+      if (shouldRethrow) {
+        kj::throwFatalException(kj::mv(*exception));
+      } else {
+        // An exception is already unwinding the stack, so throwing this secondary exception would
+        // call std::terminate().
```

Comment:
Would using C++ nested exceptions or adding extra machinery into KJ for that be helpful?

Reply:
Sadly, no.

C++ nested exceptions are not intended for this. They are intended to express an exception that is "caused by" another exception. You have to have first _caught_ the "cause" exception, and then you can construct a new exception which has the "cause" as a nested exception. Here, we've detected that an exception is in flight, but it hasn't been caught yet.

I would love for KJ to find a way to magically resolve the throw-during-destructor-during-unwind problem, as the C++ default behavior is awful. But my research indicates that the only way to do this is to modify the C++ exception machinery itself, i.e. change libgcc or whatever implements the C++ ABI. This is a bit hard for a library to do.